include/cufinufft/impl.h

```diff
@@ -196,6 +197,10 @@ int cufinufft_makeplan_impl(int type, int dim, int *nmodes, int iflag, int ntran
     printf("[cufinufft] shared memory required for the spreader: %ld\n", mem_required);
   }
 
+  // Default to method 1 for type 3.
+  if (type == 3 && d_plan->opts.gpu_method == 0)
+      d_plan->opts.gpu_method = 1;
```

Comment:
can you add brackets here. It is good practice

Reply:
The bigger question is what happens if type 3 is invoked with gpu_method=0.   Ideally we'd want the type-1-spread-only to use gpu_meth=2 (SM) iff RAM is available, but the inner-type-2 to be always meth=1 (GM sort).   Could someone check whether what happens now is that meth=1 is used in *both* the above steps?

Reply:
I'm realising we should really fix the logic here, otherwise we miss out on any way the user can access the faster type1-spreadonly for type 3.

Let's consider what happens when type-3 is done.  There is no way to have a single gpu_method control both the two subsets (t1-spreadonly, and inner-t2), and I don't want to create another opts field! So we just have to pick sensible defaults.

I propose that A) gpu_method for the t3 should be inherited by the t1-spreadonly, and that B) the inner-t2 should always be gpu_method=1 (the only option).

During setpts, B) is enforced already: l.798 currently sets the inner-t2 method to 1 (a line which is in fact redundant because the creation of t2_plan with default gpu_method=0 would select method=1 anyway). So let's remove l.798 and rely on l.221 to do its job. This is a single-source-of-truth.

Now, A) is fixed by simply moving l.213-237 up out of the t1+2 conditional block, ie, just above l.199.
Then t3 would cause the RAM-check to set the method for the t1-spreadonly, just like for a t1.

Do you both agree that this is the actual logic we want?

Reply:
PS Line numbers were from master not the proposal.

Reply:
> We don't enforce this anywhere else :) I think it's already clear for a simple Boolean like this...

Fine with me. It's good practice.

>  So let's remove l.798 and rely on l.221 to do its job. This is a single-source-of-truth.

We can't really remove it since it copies the opts from the outer plan. Set it to zero instead.

> Now, A) is fixed by simply moving l.213-237 up out of the t1+2 conditional block, ie, just above l.199.

Done.